plugins/mac-capture/window-utils.m

```diff
@@ -37,186 +36,198 @@
 #define WAIT_TIME_US WAIT_TIME_MS * 1000
 #define WAIT_TIME_NS WAIT_TIME_US * 1000
 
-bool find_window(cocoa_window_t cw, obs_data_t *settings, bool force)
+bool find_window(cocoa_window_t cw, obs_data_t *settings)
 {
-	if (!force && cw->next_search_time > os_gettime_ns())
+	uint64_t ts = os_gettime_ns();
+
+	if (cw->last_search_time + WAIT_TIME_NS > ts)
 		return false;
 
-	cw->next_search_time = os_gettime_ns() + WAIT_TIME_NS;
+	cw->last_search_time = ts;
 
-	pthread_mutex_lock(&cw->name_lock);
+	assert(pthread_mutex_lock(&cw->mutex) == 0);
 
 	if (!cw->window_name.length && !cw->owner_name.length)
-		goto invalid_name;
+		goto unlock;
 
 	for (NSDictionary *dict in enumerate_windows()) {
```

Comment:
The more I think about it, the more I wonder why these mutexes even exist in the code, as each thread has its own private structure for each Window Capture instance, so these mutexes are wholly unnecessary.  I'm going to remove them and grind this under the debugger and see what happens.

Reply:
Okay, we do need the mutex for `cocoa_window` struct, because the main thread calls `update_window()` while our capture thread calls `find_window()` and we need to make sure the values don't change out from underneath us.

Reply:
> Okay, we do need the mutex for `cocoa_window` struct, because the main thread calls `update_window()` while our capture thread calls `find_window()` and we need to make sure the values don't change out from underneath us.

We have similar mutexes for Display Capture and Window Capture on Windows for the same reason.

https://github.com/obsproject/obs-studio/blob/bf27941f5f04aa52750b34c9a204f4e7bab9bf5d/plugins/win-capture/duplicator-monitor-capture.c#L179-L182
https://github.com/obsproject/obs-studio/blob/3706a21327378eab0bd04553f3c44ac506e53e83/plugins/win-capture/window-capture.c#L186-L188

> Would it make sense to call enumerate_windows before we lock the mutex, to reduce the work done in the critical section?

I think @MakersF is suggesting something like this?
```
NSArray *arr = enumerate_windows();

pthread_mutex_lock(&cw->name_lock);

if (!cw->window_name.length && !cw->owner_name.length)
	goto invalid_name;

for (NSDictionary *dict in arr) {
	// stuff
}
```

I don't know if there's a reason we call `enumerate_windows()` directly in the `for` loop expression instead of doing it this way, and I am not familiar with Objective-C or macOS development.

Reply:
This thread is still unresolved. @MakersF @dossy ?

Reply:
Yes, I was suggesting what @RytoEX wrote. Either way, this is not a correctness issue, so it's good to go. If we are confident that there is no issue with moving out of the mutex I'd do it, but I'd rather see this PR merged rather than argue on what is best to do in this scenario, I don't think it's critical for performances